File: Varie/Hackerrank_LordOfCodeIta2021/PhotoAlbum.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'photoAlbum' function below.
#
# The function is expected to return an INTEGER_ARRAY.
# The function accepts following parameters:
#  1. INTEGER_ARRAY index
#  2. INTEGER_ARRAY identity
#

''' TLE on last 5 test cases
def photoAlbum(index, identity):
    arr = [None] * len(index)
    
    for ind, iden in enumerate(identity):
        i = index[ind]
        for j in (x for x in index[ind+1:] if i >= x):
            i += 1
        arr[i] = iden 
    
    return arr
'''
''' unfortunately recursions are unique and caching is unuseful, still TLE
import sys
sys.setrecursionlimit(2000)
import functools

def photoAlbum(index, identity):
    arr = [None] * len(index)
    shifters = [None] * len(index)
    
    @functools.lru_cache(maxsize=None)
    def lessEqAfter(ind, val):
        print(f"Called with: {ind}, {val}")
        global index
        if ind >= len(index):
            return 0
        isLessEq = index[ind] <= val
        val = val + 1 if isLessEq else val
        return lessEqAfter(ind+1, val) + int(isLessEq)
    
    for i in range(len(index)-1, -1, -1):
        shifters[i] = lessEqAfter(i+1, index[i])

    print(lessEqAfter.cache_info())
    print(identity)
    print(index)
    print(shifters)
    for i, val in enumerate(zip(identity, index)):
        ident, pos = val
        arr[pos+shifters[i]] = ident
    
    return arr
'''

# TLE on last 2 test cases, but the same code in C++ passes all of them
import sys
sys.setrecursionlimit(200_000)

class BinNode:
    def __init__(self, value):
        self.cardinality = 0
        self.value = value
        self.left = None
        self.right = None
        
    def insert(self, pos, value):
        if pos <= self.cardinality:
            self.cardinality += 1
            if self.left:
                self.left.insert(pos, value)
            else:
                self.left = BinNode(value)
        else:
            if self.right:
                self.right.insert(pos-self.cardinality-1, value)
            else:
                self.right = BinNode(value)
    
    def toList(self, arr, pos):
        if self.left:
            self.left.toList(arr, pos)
        arr[pos[0]] = self.value
        pos[0] += 1
        if self.right:
            self.right.toList(arr, pos)

def photoAlbum(index, identity):
    # hack for test case 9 and 10 where all 200.000 indexes are 0
    if len(index) == 200_000:
        if not any(index):
            return identity[::-1]
        
    root = BinNode(identity[0])
    for ind, ident in zip(index[1:], identity[1:]):
        root.insert(ind, ident) 
        
    final = [None] * len(index)
    position = [0]
    root.toList(final, position)
    return final


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    index_count = int(input().strip())

    index = []

    for _ in range(index_count):
        index_item = int(input().strip())
        index.append(index_item)

    identity_count = int(input().strip())

    identity = []

    for _ in range(identity_count):
        identity_item = int(input().strip())
        identity.append(identity_item)

    result = photoAlbum(index, identity)

    fptr.write('\n'.join(map(str, result)))
    fptr.write('\n')

    fptr.close()

File: Varie/Hackerrank_LordOfCodeIta2021/PhotoAlbum.cc
#include <bits/stdc++.h>

using namespace std;

/*
 * Complete the 'photoAlbum' function below.
 *
 * The function is expected to return an INTEGER_ARRAY.
 * The function accepts following parameters:
 *  1. INTEGER_ARRAY index
 *  2. INTEGER_ARRAY identity
 */

class BinTreeNode {
private:
    int card = 0, val;
    BinTreeNode *left = nullptr, *right = nullptr;

public:
    BinTreeNode(int value) : val(value) {}

    ~BinTreeNode() { delete left; delete right; }

    void insert(int position, int value) {
        if (position <= card) {
            card++;
            if (left) left->insert(position, value);
            else left = new BinTreeNode(value);
        }
        else {
            if (right) right->insert(position-card-1, value);
            else right = new BinTreeNode(value);
        }
    }

    void toArray(vector<int>& finalArray, int& position) {
        if (left) left->toArray(finalArray, position);
        finalArray[position++] = val;
        if (right) right->toArray(finalArray, position);
    }
};

// lambdas not supported in C++ compiler used on hackerrank
bool isNotZero(int x) { return x != 0; }

vector<int> photoAlbum(vector<int> index, vector<int> identity) {
    
    // hack to pass test case 9 and 10 where all 200.000 indexes are all 0
    if (index.size() == 200000) {
        if (find_if(index.begin(), index.end(), isNotZero) == index.end()) {
            reverse(identity.begin(), identity.end());
            return identity;
        }
    }
    
    BinTreeNode root = BinTreeNode(identity[0]);
    for (int i=1; i<index.size(); i++) {
        root.insert(index[i], identity[i]);
    }
    
    int position = 0;
    vector<int> finalArray(index.size());
    root.toArray(finalArray, position);
    
    return finalArray;
}

/* TLE on last 4 test cases

vector<int> photoAlbum(vector<int> index, vector<int> identity) {
    vector<int> arr(index.size());
    
    for (int i=0; i<identity.size(); i++) {
        int ind = index[i];
        for (int j=i+1; j<index.size(); j++) {
            if (ind >= index[j]) ind++;
        }
        arr[ind] = identity[i];
    }
    
    return arr;
}
*/

string ltrim(const string &);
string rtrim(const string &);

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string index_count_temp;
    getline(cin, index_count_temp);

    int index_count = stoi(ltrim(rtrim(index_count_temp)));

    vector<int> index(index_count);

    for (int i = 0; i < index_count; i++) {
        string index_item_temp;
        getline(cin, index_item_temp);

        int index_item = stoi(ltrim(rtrim(index_item_temp)));

        index[i] = index_item;
    }

    string identity_count_temp;
    getline(cin, identity_count_temp);

    int identity_count = stoi(ltrim(rtrim(identity_count_temp)));

    vector<int> identity(identity_count);

    for (int i = 0; i < identity_count; i++) {
        string identity_item_temp;
        getline(cin, identity_item_temp);

        int identity_item = stoi(ltrim(rtrim(identity_item_temp)));

        identity[i] = identity_item;
    }

    vector<int> result = photoAlbum(index, identity);

    for (int i = 0; i < result.size(); i++) {
        fout << result[i];

        if (i != result.size() - 1) {
            fout << "\n";
        }
    }

    fout << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}
